### PR TITLE
Add mutation for refreshing external forms

### PR DESCRIPTION
**Pull request by martha**
[//]: # 'remove this if PR is for a release-* branch'
# _Please squash merge this PR_

## Description

GH issue: https://github.com/open-path/Green-River/issues/6621
This PR adds a mutation to refresh external form submissions.

## Type of change
[//]: # 'remove options that are not relevant'
- [ ] Bug fix
- [x] New feature (adds functionality)
- [ ] Code clean-up / housekeeping
- [ ] Dependency update

## Checklist before requesting review
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] My code includes comments and/or descriptive variable names to help other engineers understand the intent (or not applicable)
- [x] My code follows the style guidelines of this project (rubocop)
- [x] I have updated the documentation (or not applicable)
- [x] If it's not obvious how to test this change, I have provided testing instructions in this PR or the related issue

### CODE REVIEW

File: drivers/hmis/app/graphql/mutations/refresh_external_submissions.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+module Mutations
+  class RefreshExternalSubmissions < CleanBaseMutation
+    field :success, Boolean, null: false
+
+    def resolve
+      handlers = ['HmisExternalApis::ConsumeExternalFormSubmissionsJob']
+      return if Delayed::Job.queued?(handlers) || Delayed::Job.running?(handlers)
+
+      queue = ENV.fetch('DJ_SHORT_QUEUE_NAME', :short_running)
+      HmisExternalApis::ConsumeExternalFormSubmissionsJob.delay(priority: 12, queue: queue)
```
Comment (martha) [inline drivers/hmis/app/graphql/mutations/refresh_external_submissions.rb]: What should priority be here? I just copied this from the service history queuer but maybe this should be higher priority?

File: drivers/hmis/app/graphql/mutations/refresh_external_submissions.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
+#  Copyright 2016 - 2024 Green River Data Analysis, LLC
+#
+#  License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+#
+
+module Mutations
```
Comment (martha) [inline drivers/hmis/app/graphql/mutations/refresh_external_submissions.rb:6]: Should this indeed be a mutation (because it is, sorta kinda, mutating data on the backend) or is it more like a query?

File: drivers/hmis/app/graphql/mutations/refresh_external_submissions.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    field :success, Boolean, null: false
+
+    def resolve
+      handlers = ['HmisExternalApis::ConsumeExternalFormSubmissionsJob']
+      return if Delayed::Job.queued?(handlers) || Delayed::Job.running?(handlers)
+
+      queue = ENV.fetch('DJ_SHORT_QUEUE_NAME', :short_running)
+      HmisExternalApis::ConsumeExternalFormSubmissionsJob.delay(priority: 12, queue: queue)
+
+      { success: true }
```
Comment (martha) [inline drivers/hmis/app/graphql/mutations/refresh_external_submissions.rb:17]: I think the ux we’re proposing is that a user hits ‘refresh’, the button spins for a second, then that query resolves, but no new records are added to the table, because a delayed job was kicked off. Then, a few seconds/minutes later, the user would need to refresh the page to actually see the new records.

I'm wondering whether this query should *also* accept a form definition ID arg and return the latest results? That way at least the user would be able to hit refresh over and over and eventually see some new data. Maybe I misunderstood the plan for the UI of this feature, though - let's discuss more.

File: drivers/hmis/app/graphql/mutations/refresh_external_submissions.rb (hunk 1 above)
Comment (martha) [reply drivers/hmis/app/graphql/mutations/refresh_external_submissions.rb]: Ah hm maybe priority should be nil/not passed, if priority is an alternative to queue as described here: https://github.com/collectiveidea/delayed_job?tab=readme-ov-file#parameters

Comment (gigxz) [reply drivers/hmis/app/graphql/mutations/refresh_external_submissions.rb]: I think it should be high priority (`priority: 0`) since it was a user-initiated action

File: drivers/hmis/app/graphql/mutations/refresh_external_submissions.rb (hunk 2 above)
Comment (gigxz) [reply drivers/hmis/app/graphql/mutations/refresh_external_submissions.rb:6]: Yeah I think mutation since it's not returning any data. It is modifying data on the server by queuing up the fetch job.

File: drivers/hmis/app/graphql/mutations/refresh_external_submissions.rb (hunk 3 above)
Comment (gigxz) [reply drivers/hmis/app/graphql/mutations/refresh_external_submissions.rb:17]: Another option would be for the frontend to invalidate the cached submissions when this mutation returns successfully, to trigger a re-fetch of the table. That way, if the job _did_ get queued and executed immediately, they would see those results. It's still not ideal but might be an easy way to improve the experience slightly.

Comment (gigxz) [reply drivers/hmis/app/graphql/mutations/refresh_external_submissions.rb:17]: Sounds like with the long-running queue it would take a few minutes for it to get queued, so it probably _won't_ be ready immediately. That said, if they click the button over and over, eventually the cache-clearing will yield results. So I think we should go with that.

File: drivers/hmis/app/graphql/mutations/refresh_external_submissions.rb [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+#  License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+#
+
+module Mutations
+  class RefreshExternalSubmissions < CleanBaseMutation
+    field :success, Boolean, null: false
+
+    def resolve
+      handlers = ['HmisExternalApis::ConsumeExternalFormSubmissionsJob']
+      return if Delayed::Job.queued?(handlers) || Delayed::Job.running?(handlers)
```
Comment (gigxz) [inline drivers/hmis/app/graphql/mutations/refresh_external_submissions.rb]: This should probably return ` { success: true }`?

File: drivers/hmis/app/graphql/mutations/refresh_external_submissions.rb (hunk 1 above)
Comment (martha) [reply drivers/hmis/app/graphql/mutations/refresh_external_submissions.rb]: It looks like we also have negative numbers for even-higher priority https://github.com/greenriver/hmis-warehouse/blob/f039ec719577065e8d3c7c0987b0ee3e3a131c40/config/initializers/delayed_job.rb#L9-L11

File: app/models/aws_s3.rb [hunk 5]
```diff
@@ -87,7 +87,7 @@ def fetch_key_list(prefix: '')
     @bucket.objects(prefix: prefix).sort_by(&:last_modified).map(&:key)
   end
 
-  def list_objects(max_keys = 1_000, prefix: '')
+  def list_objects(max_keys = 10_000, prefix: '')
```
Comment (gigxz) [inline app/models/aws_s3.rb]: I don't think we should change the default for everyone, that might have some wide-ranging impacts that we are unaware of. We can pass the `max_keys` argument to list_objects from ConsumeExternalFormSubmissionsJob instead.

File: drivers/hmis/app/graphql/mutations/refresh_external_submissions.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+module Mutations
+  class RefreshExternalSubmissions < CleanBaseMutation
+    field :success, Boolean, null: false
+
+    def resolve
+      handlers = ['HmisExternalApis::ConsumeExternalFormSubmissionsJob']
+      return { success: true } if Delayed::Job.queued?(handlers) || Delayed::Job.running?(handlers)
+
+      queue = ENV.fetch('DJ_LONG_QUEUE_NAME', :long_running)
+      HmisExternalApis::ConsumeExternalFormSubmissionsJob.delay(priority: 0, queue: queue)
```
Comment (martha) [inline drivers/hmis/app/graphql/mutations/refresh_external_submissions.rb]: Still to do: Modify the s3 client with the behavior we agreed on (https://github.com/open-path/Green-River/issues/6621#issuecomment-2368996651). I'd like to do this in a separate PR if that's ok.

**Review by gigxz — APPROVED** [review]
LGTM after isolating the max_keys change the the external forms job

File: app/models/aws_s3.rb (hunk 5 above)
Comment (martha) [reply app/models/aws_s3.rb]: Oh, thanks good catch. It looks like the s3 api won't allow a max keys of 10,000 anyway.

Comment (gigxz) [reply app/models/aws_s3.rb]: ahh k good call to revert

**Review by gigxz — APPROVED** [review]

File: drivers/hmis/app/graphql/mutations/refresh_external_submissions.rb (hunk 6 above)
Comment (martha) [reply drivers/hmis/app/graphql/mutations/refresh_external_submissions.rb]: Edit: This is split out into a separate ticket https://github.com/open-path/Green-River/issues/6727
